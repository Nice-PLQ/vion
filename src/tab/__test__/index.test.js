import { mount } from '@vue/test-utils';
import { next, triggerDrag } from '../../../test';
import VTab from '../tab.vue';
import VTabPanel from '../../tab-panel';

describe('Tab', () => {
  const consoleOutput = [];
  beforeEach(() => {
    console.error = output => consoleOutput.push(output);
    console.log = output => consoleOutput.push(output);
  });
  const wrapper = mount({
    template: `
      <v-tab>
        <v-tab-panel v-for="(tab,index) in tabs" :title="tab" :key="index">{{tab}}</v-tab-panel>
      </v-tab>
    `,
    data() {
      return {
        tabs: [
          '热门赛事',
          'QQ飞车手游',
          '英雄联盟',
          '王者荣耀',
          '和平精英'
        ]
      };
    }
  }, {
    global: {
      components: { VTab, VTabPanel }
    }
  });

  test('渲染', async () => {
    const tabItem = wrapper.findAll('.v-tab__item')[2];
    await tabItem.trigger('click');
    expect(tabItem.classes()).toContain('on');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('tab拖动', () => {
    const el = wrapper.find('.v-tab__nav');
    triggerDrag(el, 50, 0);
    expect(el.element.style.webkitTransition).toMatch(/cubic-bezier/);
  });

  test('tab数据格式', async () => {
    const fn = jest.fn();
    const wrapper = mount({
      template: `
        <v-tab @change="onChange" v-model="active">
          <v-tab-panel v-for="(tab,index) in tabs" :title="tab.title" :name="tab.name" :key="index">{{tab.title}}</v-tab-panel>
        </v-tab>
      `,
      data() {
        return {
          tabs: [
            { title: '好友', name: 'a' },
            { title: '分组', name: 'b' },
            { title: '群聊', name: 'c' }
          ],
          active: 'a'
        };
      },
      methods: {
        onChange: fn
      }
    }, {
      global: {
        components: { VTab, VTabPanel }
      }
    });

    await next();
    const tabItem = wrapper.findAll('.v-tab__item')[2];
    await tabItem.trigger('click');
    expect(fn).toHaveBeenCalled();
    expect(wrapper.element).toMatchSnapshot();
  });

  test('name属性重复校验', async () => {
    const wrapper = mount({
      template: `
        <v-tab>
          <v-tab-panel v-for="(tab,index) in tabs" :title="tab.title" :name="tab.name" :key="index">{{tab.title}}</v-tab-panel>
        </v-tab>
      `,
      data() {
        return {
          tabs: [
            { title: '好友', name: 'a' },
            { title: '分组', name: 'a' }
          ]
        };
      }
    }, {
      global: {
        components: { VTab, VTabPanel }
      }
    });
    expect(wrapper.element).toMatchSnapshot();
    expect(consoleOutput).toContain("duplicated 'name' property at v-tab-panel component");
  });
});
