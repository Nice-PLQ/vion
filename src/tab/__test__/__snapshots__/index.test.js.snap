// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab name属性重复校验 1`] = `
<div
  class="v-tab"
>
  <div
    class="v-tab__nav-wrap"
  >
    <ul
      class="v-tab__nav"
      role="tablist"
      style="transform: translate3d(0px, 0, 0);"
    >
      
      
    </ul>
  </div>
  <div
    class="v-tab__content"
    style="transform: translate3d(0px, 0, 0); transition: all 400ms ease;"
  >
    
    
    <div
      class="v-tab__panel"
    >
      
      好友
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      分组
      
    </div>
    
    
  </div>
</div>
`;

exports[`Tab tab数据格式 1`] = `
<div
  class="v-tab"
>
  <div
    class="v-tab__nav-wrap"
  >
    <ul
      class="v-tab__nav"
      role="tablist"
      style="transform: translate3d(0px, 0, 0);"
    >
      
      <li
        class="v-tab__item"
        role="tab"
      >
        好友
      </li>
      <li
        class="v-tab__item"
        role="tab"
      >
        分组
      </li>
      <li
        aria-selected="true"
        class="v-tab__item on"
        role="tab"
      >
        群聊
      </li>
      
    </ul>
  </div>
  <div
    class="v-tab__content"
    style="transform: translate3d(0px, 0, 0); transition: all 400ms ease;"
  >
    
    
    <div
      class="v-tab__panel"
    >
      
      好友
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      分组
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      群聊
      
    </div>
    
    
  </div>
</div>
`;

exports[`Tab 渲染 1`] = `
<div
  class="v-tab"
>
  <div
    class="v-tab__nav-wrap"
  >
    <ul
      class="v-tab__nav v-tab__nav_scrollable"
      role="tablist"
      style="transform: translate3d(0px, 0, 0); webkit-transition: all 400ms ease;"
    >
      
      <li
        class="v-tab__item"
        role="tab"
      >
        热门赛事
      </li>
      <li
        class="v-tab__item"
        role="tab"
      >
        QQ飞车手游
      </li>
      <li
        aria-selected="true"
        class="v-tab__item on"
        role="tab"
      >
        英雄联盟
      </li>
      <li
        class="v-tab__item"
        role="tab"
      >
        王者荣耀
      </li>
      <li
        class="v-tab__item"
        role="tab"
      >
        和平精英
      </li>
      
    </ul>
  </div>
  <div
    class="v-tab__content"
    style="transform: translate3d(0px, 0, 0); transition: all 400ms ease;"
  >
    
    
    <div
      class="v-tab__panel"
    >
      
      热门赛事
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      QQ飞车手游
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      英雄联盟
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      王者荣耀
      
    </div>
    <div
      class="v-tab__panel"
    >
      
      和平精英
      
    </div>
    
    
  </div>
</div>
`;
